0003-interledger-protocol/0003-interledger-protocol.md

```diff
@@ -289,6 +289,7 @@ Final errors indicate that the payment is invalid and should not be retried unle
 | **F05** | **Wrong Condition** | The receiver generated a different condition and cannot fulfill the payment. |
 | **F06** | **Unexpected Payment** | The receiver was not expecting a payment like this (the memo and destination address don't make sense in that combination, for example if the receiver does not understand the transport protocol used) |
 | **F07** | **Cannot Receive** | The receiver is unable to accept this payment due to a constraint. For example, the payment would put the receiver above its maximum account balance. |
+| **F08** | **Amount Too Large** | The amount is larger than the maximum size supported by a connector, ledger or the receiver. The producer of this error SHOULD set the error `data` to the concatenation of the following two numbers, each encoded as a 64-bit unsigned, big-endian integer: the amount that arrived and the maximum amount supported by the party producing the error. |
```

Comment:
Couple of questions:

1. It's perhaps a distinction without a difference, but specifying the encoding of the data field here feels wrong to me. Shouldn't this be left to the ASN.1 definition, and be more loosely specified here?

1. Should those amounts be defined at all here in the spec? In the case of `F03` and `F04`, we don't specify the amounts that were too low. Seems like we should have consistency, and either mandate amounts in those errors too, or remove the amounts from this (F08) error.

1. Does this use-case (i.e., wanting to know specific and additional information about amounts in an ILP error) indicate that the `data` field, as defined, is insufficient for ILP errors? Instead, should we remove the `data` field and move to something like the `Subprotocol` concept? In that world, a subprotocol could define what information to provide (e.g., the amount that arrived and the max amount supported by the party) _and_ how to encode it.

1. Outside of sub-protocol, there are certain errors that require additional properties. Would it make sense to define actual error packets instead? Or use a collection of key-value pairs that could be easily extensible?

Overall, it feels like a bolt-on to just stuff two numbers into the `data` packet in such a specific manner - my feeling is that we're going to need the ability to include extra information about various errors in the future, and we should have a well-defined way to do such a thing in a consistent manner.

Not sure what that "consistent manner" should be just yet though...

Reply:
Fair points.

**Should errors carry structured data?** Probably. There are at least a couple of cases where it will make the handling of those errors significantly more efficient.

**Is the extra data part of an extension or subprotocol, or is it linked to the error?** I think the more common case will be extra data that is intimately linked to the error. End-to-end protocols can include whatever data they want in the error data field already, so the case we're really talking about is errors that are produced and/or understood by senders and connectors.

**Should the error format include a place to put key-value pairs?** If the extra data that needs to be included is linked to the error itself, I don't think it needs to have the keys and could be encoded as just the values. Keys are most useful to include for extensibility mechanisms where you don't know beforehand which values will be included. However, if the data is thought of as part of an individual error, we do know the fields that will be there up front. If we use OER encoding, a `SEQUENCE` would just be the values of the individual fields concatenated together (I was thinking about that when I suggested that the two numbers be concatenated in the data field).

**Should the general error format include a place to put additional error-specific fields or should that be left to the definition of individual errors?** If there are fields that will be common across many different errors it would make sense to add those to the general error format itself. If the extra data is error-specific, I think it would make sense to have a generic field like `data` in the general format, and then to have specific errors dictate what should go in that field.

So my thoughts are that:
- we should keep the general error format as it is
- individual errors can specify what additional information SHOULD be put in the `data` field
- additional data SHOULD be encoded using OER
- the definitions of the additional data should be wherever the specific errors are defined (right now they're all in RFC-3), and we could additionally put them in an ASN.1 file